File: src/python/_0002_Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1, l2):
        first = temp = ListNode(0)
        carry = 0

        while l1 or l2 or carry:
            carry += l1.val if l1 else 0
            carry += l2.val if l2 else 0

            temp.next = ListNode(carry % 10)
            temp = temp.next

            carry //= 10

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        return first.next

File: src/_0002_Add_Two_Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode first = new ListNode(0), curr = first;
        int carry = 0;

        while (l1 != null || l2 != null || carry != 0) {
            carry += l1 != null ? l1.val : 0;
            carry += l2 != null ? l2.val : 0;

            curr.next = new ListNode(carry % 10);
            curr = curr.next;

            carry /= 10;

            l1 = l1 != null ? l1.next : null;
            l2 = l2 != null ? l2.next : null;
        }

        return first.next;
    }
}
